src/libopenrave-core/jsonparser/jsonreader.cpp

```diff
@@ -298,6 +298,36 @@ class JSONReader
                 }
                 return false;
             }
+            // if parent rEnvInfo is going to overwrite bodies in the environment of the referenceUri, do not deserialize them. otherwise, they would be merged instead of overwritten.
+            if( rEnvInfo.HasMember("bodies") && rEnvInfo["bodies"].Size() > 0 ) {
+                if( prReferenceEnvInfo->HasMember("bodies") && (*prReferenceEnvInfo)["bodies"].Size() > 0 ) {
+                    boost::shared_ptr<rapidjson::Document> prFilteredReferenceEnvInfo = boost::make_shared<rapidjson::Document>(&alloc);
+                    prFilteredReferenceEnvInfo->CopyFrom(*prReferenceEnvInfo, alloc);
+                    const rapidjson::Value& rEnvBodies = rEnvInfo["bodies"];
+                    rapidjson::Value& rRefBodies = (*prFilteredReferenceEnvInfo)["bodies"];
+                    for (rapidjson::Value::ConstValueIterator itEnvBody = rEnvBodies.Begin(); itEnvBody != rEnvBodies.End(); ++itEnvBody) {
+                        rapidjson::Value::ConstMemberIterator itEnvBodyId = itEnvBody->FindMember("id");
+                        if( itEnvBodyId == itEnvBody->MemberEnd() || itEnvBodyId->value.GetStringLength() == 0 ) {
+                            continue;
+                        }
+                        const char* envBodyId = itEnvBodyId->value.GetString();
+                        const size_t envBodyIdLen = itEnvBodyId->value.GetStringLength();
+                        for (rapidjson::Value::ValueIterator itRefBody = rRefBodies.Begin(); itRefBody != rRefBodies.End(); ++itRefBody) {
+                            rapidjson::Value::ConstMemberIterator itRefBodyId = itRefBody->FindMember("id");
+                            if( itRefBodyId == itRefBody->MemberEnd() || itRefBodyId->value.GetStringLength() == 0 ) {
+                                continue;
+                            }
+                            const char* refBodyId = itRefBodyId->value.GetString();
+                            if( strncmp(envBodyId, refBodyId, envBodyIdLen) == 0 ) {
```

Comment:
I think this comparison will have false positives:
```
const char* envBodyId = itEnvBodyId->value.GetString(); // let's assume "myBody"
const size_t envBodyIdLen = itEnvBodyId->value.GetStringLength(); // len("myBody")
...
const char* refBodyId = itRefBodyId->value.GetString(); // let's assume "myBody2"
strncmp(envBodyId, refBodyId, envBodyIdLen) -> strncmp("myBody, "myBody2", strlen("myBody")) -> strings are equal!
```
need to check if lengths match, or just wrap both in string_views and use the compare operator on the string view instead of manually doing strncmp.

Reply:
oops. thanks. fixed.